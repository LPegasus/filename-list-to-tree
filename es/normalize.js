export function normalizePath(pathname) {
    let path = pathname;
    if (path.indexOf("\\") > 0) {
        path = pathname.replace(/\\/g, "/");
    }
    path = path.replace(/\/$/, "");
    if (path === ".") {
        return "";
    }
    const xpath = path.split("/");
    let i = 0;
    const remain = [];
    while (i < xpath.length) {
        const c = xpath[i];
        i++;
        if (c === "..") {
            remain.pop();
        }
        else if (c !== ".") {
            remain.push(c);
        }
    }
    return remain.join("/");
}
export function relative(from, to) {
    if (from === "") {
        return normalizePath(to);
    }
    const xpathFrom = normalizePath(from).split("/");
    const xpathTo = normalizePath(to).split("/");
    let i = 0;
    while (i < xpathFrom.length) {
        const p1 = xpathFrom[i];
        const p2 = xpathTo[i];
        if (p1 !== p2) {
            break;
        }
        i++;
    }
    const rest = xpathFrom.length - i;
    const prefix = "../".repeat(rest);
    return prefix + xpathTo.slice(i).join("/");
}
