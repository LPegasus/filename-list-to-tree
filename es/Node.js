import { normalizePath, relative } from "./normalize";
export class Node {
    constructor(absolutePath, parent = null) {
        /** If null, means the node is root */
        this.parent = null;
        /** If null, means this is a file. Other wise, this is a directory. */
        this.filesOrFolders = null;
        this.absolutePath = normalizePath(absolutePath);
        if (parent) {
            this.updateParent(parent);
        }
    }
    static From(files) {
        return _generateRootNode(files);
    }
    get isTheOnlyChildOfParent() {
        const parent = this.parent;
        if (!parent) {
            return false;
        }
        const f = parent.filesOrFolders;
        return f.length === 1 && f[0] === this;
    }
    get isFolder() {
        return this.filesOrFolders !== null;
    }
    get isFile() {
        return !this.isFolder;
    }
    _addFileOrFolder(...nodeList) {
        if (!this.filesOrFolders) {
            this.filesOrFolders = [];
        }
        nodeList.forEach((d) => {
            if (!this.filesOrFolders.includes(d)) {
                this.filesOrFolders.push(d);
            }
        });
        return this;
    }
    get relativePath() {
        return relative(this.parent?.absolutePath ?? "", this.absolutePath);
    }
    detachFromParent() {
        if (this.parent) {
            const i = this.parent.filesOrFolders.indexOf(this);
            this.parent.filesOrFolders.splice(i, 1);
            this.parent = null;
        }
        return this;
    }
    updateParent(node) {
        if (node === this.parent) {
            return false;
        }
        this.detachFromParent();
        this.parent = node;
        node._addFileOrFolder(this);
        return true;
    }
    combinePathAncestors() {
        let prt = this.parent;
        const prtToDetach = [];
        while (prt.isTheOnlyChildOfParent) {
            prt = prt.parent;
            prtToDetach.push(prt);
        }
        prtToDetach.forEach((d) => d.detach());
        return this;
    }
    toJSON() {
        const { absolutePath, isFile, isFolder, filesOrFolders } = this;
        return {
            relativePath: this.relativePath,
            absolutePath,
            isFile,
            isFolder,
            filesOrFolders: filesOrFolders?.map((d) => d.toJSON()),
        };
    }
    combinePath() {
        const allFiles = [];
        const addFile = (d) => {
            if (d.isFile) {
                allFiles.push(d);
                return;
            }
            d.filesOrFolders?.forEach(addFile);
        };
        this.filesOrFolders.forEach(addFile);
        allFiles.forEach((d) => {
            d.combinePathAncestors();
        });
        return this;
    }
    detach() {
        const p = this.parent;
        this.parent = null;
        if (p) {
            const i = p.filesOrFolders.findIndex((d) => d === this);
            if (i >= 0) {
                p.filesOrFolders.splice(i, 1);
            }
        }
        this.filesOrFolders?.forEach((d) => d.updateParent(p));
        this.filesOrFolders = null;
    }
}
function _generateRootNode(files) {
    const allFiles = files.map((d) => normalizePath(d));
    const root = new Node("");
    const nodeByPath = { "": root };
    allFiles.forEach((fileName) => {
        const xpath = fileName.split("/");
        let i = 0;
        let parent = root;
        while (i < xpath.length) {
            parent = getOrCreateFolderNode(xpath.slice(0, ++i).join("/"), parent);
        }
    });
    function getOrCreateFolderNode(p, parent) {
        if (!nodeByPath[p]) {
            nodeByPath[p] = new Node(p);
        }
        nodeByPath[p].updateParent(nodeByPath[parent.absolutePath]);
        return nodeByPath[p];
    }
    return root;
}
