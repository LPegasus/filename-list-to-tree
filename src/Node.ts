import { normalizePath, relative } from "./normalize";

export class Node {
  public static From(files: string[]): Node {
    return _generateRootNode(files);
  }

  /** If null, means the node is root */
  public parent: null | Node = null;

  /** If null, means this is a file. Other wise, this is a directory. */
  public filesOrFolders: null | Node[] = null;

  public absolutePath: string;

  public get isTheOnlyChildOfParent(): boolean {
    const parent = this.parent;
    if (!parent) {
      return false;
    }
    const f = parent.filesOrFolders;
    return f.length === 1 && f[0] === this;
  }

  public get isFolder(): boolean {
    return this.filesOrFolders !== null;
  }

  public get isFile(): boolean {
    return !this.isFolder;
  }

  private _addFileOrFolder(...nodeList: Node[]) {
    if (!this.filesOrFolders) {
      this.filesOrFolders = [];
    }
    nodeList.forEach((d) => {
      if (!this.filesOrFolders!.includes(d)) {
        this.filesOrFolders!.push(d);
      }
    });
    return this;
  }

  public get relativePath(): string {
    return relative(this.parent?.absolutePath ?? "", this.absolutePath);
  }

  public detachFromParent() {
    if (this.parent) {
      const i = this.parent.filesOrFolders!.indexOf(this);
      this.parent.filesOrFolders!.splice(i, 1);
      this.parent = null;
    }
    return this;
  }

  public updateParent(node: Node) {
    if (node === this.parent) {
      return false;
    }
    this.detachFromParent();
    this.parent = node;
    node._addFileOrFolder(this);
    return true;
  }

  public combinePathAncestors() {
    let prt: Node = this.parent;
    const prtToDetach: Node[] = [];
    while (prt.isTheOnlyChildOfParent) {
      prt = prt.parent;
      prtToDetach.push(prt);
    }
    prtToDetach.forEach((d) => d.detach());
    return this;
  }

  public toJSON() {
    const { absolutePath, isFile, isFolder, filesOrFolders } = this;
    return {
      relativePath: this.relativePath,
      absolutePath,
      isFile,
      isFolder,
      filesOrFolders: filesOrFolders?.map((d) => d.toJSON()),
    };
  }

  public combinePath() {
    const allFiles: Node[] = [];
    const addFile = (d: Node) => {
      if (d.isFile) {
        allFiles.push(d);
        return;
      }
      d.filesOrFolders?.forEach(addFile);
    };
    this.filesOrFolders.forEach(addFile);

    allFiles.forEach((d) => {
      d.combinePathAncestors();
    });
    return this;
  }

  public detach(): void {
    const p = this.parent;
    this.parent = null;
    if (p) {
      const i = p.filesOrFolders.findIndex((d) => d === this);
      if (i >= 0) {
        p.filesOrFolders.splice(i, 1);
      }
    }
    this.filesOrFolders?.forEach((d) => d.updateParent(p));
    this.filesOrFolders = null;
  }

  constructor(absolutePath: string, parent: Node | null = null) {
    this.absolutePath = normalizePath(absolutePath);
    if (parent) {
      this.updateParent(parent);
    }
  }
}

function _generateRootNode(files: string[]) {
  const allFiles = files.map((d) => normalizePath(d));
  const root = new Node("");
  const nodeByPath: Record<string, Node> = { "": root };
  allFiles.forEach((fileName) => {
    const xpath = fileName.split("/");
    let i = 0;
    let parent = root;
    while (i < xpath.length) {
      parent = getOrCreateFolderNode(xpath.slice(0, ++i).join("/"), parent);
    }
  });

  function getOrCreateFolderNode(p: string, parent: Node) {
    if (!nodeByPath[p]) {
      nodeByPath[p] = new Node(p);
    }
    nodeByPath[p].updateParent(nodeByPath[parent.absolutePath]);
    return nodeByPath[p];
  }
  return root;
}
