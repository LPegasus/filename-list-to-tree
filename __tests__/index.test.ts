import { normalizePath, relative } from "../src/normalize";
import { Node } from "../src/Node";

it("normalizePath", () => {
  expect(normalizePath(".")).toEqual("");
  expect(normalizePath("a/b/c/")).toEqual("a/b/c");
  expect(normalizePath("a/b/c")).toEqual("a/b/c");
  expect(normalizePath("/a/b/c")).toEqual("/a/b/c");
  expect(normalizePath("a/././b/./c/./")).toEqual("a/b/c");
  expect(normalizePath("a/../b/../c/")).toEqual("c");
  expect(normalizePath("a\\b\\.\\c\\")).toEqual("a/b/c");
});

it("relative", () => {
  expect(relative("a/b/", "a/b/c/d")).toEqual("c/d");
  expect(relative("a/c/d", "a/b")).toEqual("../../b");
  expect(relative("", ".gitignore")).toEqual(".gitignore");
});

it("generateRootNode", () => {
  const allFiles: string[] = [
    ".gitignore",
    "apps/infra/libs/package.json",
    "apps/infra/project1/src/App.tsx",
    "apps/infra/project1/src/index.tsx",
    "apps/infra/project1/package.json",
    "common/config/rush/.pnpmfile.cjs",
    "common/config/rush/common-version.json",
    "package.json",
  ];
  const root = Node.From(allFiles);
  expect(root.toJSON()).toMatchSnapshot();
});

it("generateRootNode combine path", () => {
  const allFiles: string[] = [
    ".gitignore",
    "apps/tools/tool1/tool2/package.json",
    "apps/infra/libs/package.json",
    "apps/infra/project1/src/App.tsx",
    "apps/infra/project1/src/index.tsx",
    "apps/infra/project1/package.json",
    "common/config/rush/.pnpmfile.cjs",
    "common/config/rush/common-version.json",
    "package.json",
  ];
  const root = Node.From(allFiles);
  const result = root.combinePath().toJSON();
  const assertFn = (d: Node) => {
    if (d.isFolder) {
      const folders = d.filesOrFolders.filter((dd) => dd.isFolder);
      expect(d.filesOrFolders.length > 1 || folders.length === 0).toBe(true);
      folders.forEach((f) => assertFn(f));
    }
  };
  result.filesOrFolders.forEach(assertFn);
  expect(result).toMatchSnapshot();
});
