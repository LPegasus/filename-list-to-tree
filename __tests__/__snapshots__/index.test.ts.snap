// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateRootNode 1`] = `
Object {
  "absolutePath": "",
  "filesOrFolders": Array [
    Object {
      "absolutePath": ".gitignore",
      "filesOrFolders": undefined,
      "isFile": true,
      "isFolder": false,
      "relativePath": ".gitignore",
    },
    Object {
      "absolutePath": "apps",
      "filesOrFolders": Array [
        Object {
          "absolutePath": "apps/infra",
          "filesOrFolders": Array [
            Object {
              "absolutePath": "apps/infra/libs",
              "filesOrFolders": Array [
                Object {
                  "absolutePath": "apps/infra/libs/package.json",
                  "filesOrFolders": undefined,
                  "isFile": true,
                  "isFolder": false,
                  "relativePath": "package.json",
                },
              ],
              "isFile": false,
              "isFolder": true,
              "relativePath": "libs",
            },
            Object {
              "absolutePath": "apps/infra/project1",
              "filesOrFolders": Array [
                Object {
                  "absolutePath": "apps/infra/project1/src",
                  "filesOrFolders": Array [
                    Object {
                      "absolutePath": "apps/infra/project1/src/App.tsx",
                      "filesOrFolders": undefined,
                      "isFile": true,
                      "isFolder": false,
                      "relativePath": "App.tsx",
                    },
                    Object {
                      "absolutePath": "apps/infra/project1/src/index.tsx",
                      "filesOrFolders": undefined,
                      "isFile": true,
                      "isFolder": false,
                      "relativePath": "index.tsx",
                    },
                  ],
                  "isFile": false,
                  "isFolder": true,
                  "relativePath": "src",
                },
                Object {
                  "absolutePath": "apps/infra/project1/package.json",
                  "filesOrFolders": undefined,
                  "isFile": true,
                  "isFolder": false,
                  "relativePath": "package.json",
                },
              ],
              "isFile": false,
              "isFolder": true,
              "relativePath": "project1",
            },
          ],
          "isFile": false,
          "isFolder": true,
          "relativePath": "infra",
        },
      ],
      "isFile": false,
      "isFolder": true,
      "relativePath": "apps",
    },
    Object {
      "absolutePath": "common",
      "filesOrFolders": Array [
        Object {
          "absolutePath": "common/config",
          "filesOrFolders": Array [
            Object {
              "absolutePath": "common/config/rush",
              "filesOrFolders": Array [
                Object {
                  "absolutePath": "common/config/rush/.pnpmfile.cjs",
                  "filesOrFolders": undefined,
                  "isFile": true,
                  "isFolder": false,
                  "relativePath": ".pnpmfile.cjs",
                },
                Object {
                  "absolutePath": "common/config/rush/common-version.json",
                  "filesOrFolders": undefined,
                  "isFile": true,
                  "isFolder": false,
                  "relativePath": "common-version.json",
                },
              ],
              "isFile": false,
              "isFolder": true,
              "relativePath": "rush",
            },
          ],
          "isFile": false,
          "isFolder": true,
          "relativePath": "config",
        },
      ],
      "isFile": false,
      "isFolder": true,
      "relativePath": "common",
    },
    Object {
      "absolutePath": "package.json",
      "filesOrFolders": undefined,
      "isFile": true,
      "isFolder": false,
      "relativePath": "package.json",
    },
  ],
  "isFile": false,
  "isFolder": true,
  "relativePath": "",
}
`;

exports[`generateRootNode combine path 1`] = `
Object {
  "absolutePath": "",
  "filesOrFolders": Array [
    Object {
      "absolutePath": ".gitignore",
      "filesOrFolders": undefined,
      "isFile": true,
      "isFolder": false,
      "relativePath": ".gitignore",
    },
    Object {
      "absolutePath": "apps",
      "filesOrFolders": Array [
        Object {
          "absolutePath": "apps/infra",
          "filesOrFolders": Array [
            Object {
              "absolutePath": "apps/infra/libs",
              "filesOrFolders": Array [
                Object {
                  "absolutePath": "apps/infra/libs/package.json",
                  "filesOrFolders": undefined,
                  "isFile": true,
                  "isFolder": false,
                  "relativePath": "package.json",
                },
              ],
              "isFile": false,
              "isFolder": true,
              "relativePath": "libs",
            },
            Object {
              "absolutePath": "apps/infra/project1",
              "filesOrFolders": Array [
                Object {
                  "absolutePath": "apps/infra/project1/src",
                  "filesOrFolders": Array [
                    Object {
                      "absolutePath": "apps/infra/project1/src/App.tsx",
                      "filesOrFolders": undefined,
                      "isFile": true,
                      "isFolder": false,
                      "relativePath": "App.tsx",
                    },
                    Object {
                      "absolutePath": "apps/infra/project1/src/index.tsx",
                      "filesOrFolders": undefined,
                      "isFile": true,
                      "isFolder": false,
                      "relativePath": "index.tsx",
                    },
                  ],
                  "isFile": false,
                  "isFolder": true,
                  "relativePath": "src",
                },
                Object {
                  "absolutePath": "apps/infra/project1/package.json",
                  "filesOrFolders": undefined,
                  "isFile": true,
                  "isFolder": false,
                  "relativePath": "package.json",
                },
              ],
              "isFile": false,
              "isFolder": true,
              "relativePath": "project1",
            },
          ],
          "isFile": false,
          "isFolder": true,
          "relativePath": "infra",
        },
        Object {
          "absolutePath": "apps/tools/tool1/tool2",
          "filesOrFolders": Array [
            Object {
              "absolutePath": "apps/tools/tool1/tool2/package.json",
              "filesOrFolders": undefined,
              "isFile": true,
              "isFolder": false,
              "relativePath": "package.json",
            },
          ],
          "isFile": false,
          "isFolder": true,
          "relativePath": "tools/tool1/tool2",
        },
      ],
      "isFile": false,
      "isFolder": true,
      "relativePath": "apps",
    },
    Object {
      "absolutePath": "package.json",
      "filesOrFolders": undefined,
      "isFile": true,
      "isFolder": false,
      "relativePath": "package.json",
    },
    Object {
      "absolutePath": "common/config/rush",
      "filesOrFolders": Array [
        Object {
          "absolutePath": "common/config/rush/.pnpmfile.cjs",
          "filesOrFolders": undefined,
          "isFile": true,
          "isFolder": false,
          "relativePath": ".pnpmfile.cjs",
        },
        Object {
          "absolutePath": "common/config/rush/common-version.json",
          "filesOrFolders": undefined,
          "isFile": true,
          "isFolder": false,
          "relativePath": "common-version.json",
        },
      ],
      "isFile": false,
      "isFolder": true,
      "relativePath": "common/config/rush",
    },
  ],
  "isFile": false,
  "isFolder": true,
  "relativePath": "",
}
`;
